extensions/lem-dashboard/lem-dashboard.lisp

```diff
@@ -0,0 +1,128 @@
+(defpackage :lem-dashboard
+  (:use :cl :lem)
+  (:export :open-dashboard
+           :*dashboard-mode-keymap*
+           :*dashboard-enable*
+           :move-to-recent-projects
+           :move-to-recent-files
+           :set-dashboard
+           :set-default-dashboard
+           :create-centered-string))
+
+(in-package :lem-dashboard)
+
+(defvar *dashboard-buffer-name* "*dashboard*")
+(defvar *dashboard-enable* t)
+(defvar *dashboard-mode-keymap* (make-keymap :name '*dashboard-mode-keymap* :parent *global-keymap*))
+(defvar *dashboard-layout* nil
+  "List of dashboard-item instances; will be drawn in order.")
+
+(defun create-centered-string (str)
+  "Creates a 'centered' string by padding with space to fill the screen width halfway."
+  (let* ((padding (max 0 (floor (- (window-width (current-window)) (length str)) 2)))
+         (spaces (make-string padding :initial-element #\Space)))
+    (concatenate 'string spaces str)))
+
+(defclass dashboard-item ()
+  ((item-attribute 
+    :initarg :item-attribute 
+    :accessor item-attribute 
+    :initform 'document-text-attribute
+    :documentation "Attribute to use when drawing this item.")
+   (top-margin
+    :initarg :top-margin 
+    :accessor top-margin
+    :initform 0
+    :documentation "The amount of vertical space (lines) to apply before the item.")
+   (bottom-margin
+    :initarg :bottom-margin 
+    :accessor bottom-margin
+    :initform 1
+    :documentation "The amount of vertical space (lines) to apply after the item.")
+   (action 
+    :initarg :action 
+    :accessor action 
+    :initform nil
+    :documentation "Function to execute when <return> is pressed over this item."))
+  (:documentation "Base class for all dashboard items."))
+
+(defgeneric draw-dashboard-item (item point)
+  (:documentation "Called to draw the dashboard item.")
+  (:method :before ((item dashboard-item) point)
+    (dotimes (i (top-margin item))
+      (insert-character point #\Newline)))
+  (:method :after ((item dashboard-item) point)
+    (dotimes (i (bottom-margin item))
+      (insert-character point #\Newline))))
+
+(define-command dashboard-open-selected-item () ()
+  "Execute action on selected dashboard item."
+  (let* ((point (current-point))
+         (item (text-property-at point :dashboard-item)))
+    (when (and item (action item))
+      (funcall (action item)))))
+
+(defmethod draw-dashboard-item :around ((item dashboard-item) point)
+  "Inserts a :dashboard-item text property in between the starting and ending POINT, useful for tracking."
+  (let ((start (copy-point point :temporary)))
+    (call-next-method)
+    (let ((end (copy-point point :temporary)))
+      (put-text-property start end :dashboard-item item)
+      (delete-point start)
+      (delete-point end))))
```

Comment:
Can we use buttons for dashboard-item?
https://github.com/lem-project/lem/blob/main/src/ext/button.lisp

This allows for common mouse clicks, etc.

Reply:
I was gonna do that but here are some issues with making the base `dashboard-item` class a button:

- This bit sets the entire `dashboard-item` region to a button. That's fine for some items, like commands/urls (which are already buttons), but not for projects/files, since it would make the entire region a single button. We probably want each individual project/file to be its own button instead of one big one.
- Some items might not want to be buttons (e.g. splash/footer)

Reply:
OK, I got it.